apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt

```diff
@@ -44,36 +44,34 @@ class DiskLruHttpCache(private val fileSystem: FileSystem, private val directory
   }
 
   fun write(response: HttpResponse, cacheKey: String) {
-    val editor = cacheLock.read {
+    val editor = cacheLock.write {
       cache.edit(cacheKey)
```

Comment:
My understanding was that the cache could edit several (different) entries in parallel so putting all that under a write lock might increase contention?

More generally, this raises the question of what to do if two threads try to write the same cache entry concurrently. We can either decide that:

1. first one wins, second one is discarded
or
2. serialize everything. both get written to the disk

While 2. seems safer, I'm wondering if 1. is not better in terms of performance. Especially at this level of concurrency, there's no real notion of "first" and "second". Which one is going to come first is scheduling dependant so might as well be random. Writing the first one doesn't seem all that bad compared to writing both. 

All in all, I'm wondering if the bug isn't at https://github.com/apollographql/apollo-android/blob/ac5ed15dd4c0795daa0ffb089eb483845422d5fc/apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/CachingHttpInterceptor.kt#L82

Maybe it's just unsafe to assume a read after a write always succeeds and we should instead clone the network response body?

Reply:
Good analysis! I did 1. in 3b4f9ae3ce3c4baa588772efaa0478be3a0fa9a1 and it works! 🎉